### ISSUE TITLE
darks and nightly bias when only one camera is bad

### ISSUE THREAD

**Issue by sbailey** [desihub/desispec#1511]
Do `desispec.ccdcalib.make_dark_scripts`, `desi_proc --nightlybias ...`, and `desi_run_night` do the right bookkeeping if only one camera is flagged as bad instead of the full exposure?

Example: night 20211203 exposures 111915-111939 (ZERO) and 111941 (DARK) had bad z3 (VCCD=off), but were fine for the other cameras.  Observers then re-took ZEROs (111987-112011) and DARK (112013) with all 30 cameras.  For the moment I have flagged daily/exposure_tables/202112/exposure_table_20211203.csv EXPID 111915-111939 with BADCAMWORD=z3 and flagged the first DARK 111941 as bad in the hopes that pipeline re-runs will correctly pickup the second dark (112013) for determining badcolumns.  If we need to, we could flag all the first ZEROs as bad since we have a complete second set, but there may be cases in the future where we don't have a second set and need to have this bookkeeping right when a subset of cameras are bad for a subset of exposures.

@Waelthus could you check `desispec.ccdcalib.make_dark_scripts` and make updates as needed so that it doesn't incorrectly try to use the bad z3 exposures when making darks?

@Waelthus and/or @RaffaellaCapasso please try running `desi_run_night -n 20211203 --proc-obstypes dark --dry-run` and `desi_proc -n 20211203 -e 112013 --nightlybias` in a test prod to see if it incorrectly tries to use the bad z3 exposures.  We may need to get @akremin help for bookkeeping but it would be helpful if you could start with assessing the situation so that Anthony can focus on Fuji instead of dailyops.

Again: for this particular case we can flag 111915-111939 as bad if needed so it doesn't have to be a blocking factor, but it's a good test case for studying the bookkeeping for the future.

**Comment by Waelthus** [comment]
Ok, I did run the test_prod, and to me it does not look like 111915-111939 (ZERO) and 111941 (DARK) have been used for the z3 reduction. 

The processing table produced from `desi_run_night` only contains 112013 and the logs from `desi_proc` only mention that tile as well regarding badcolumns. I guess given that the full DARK was flagged to ignore, and not just Z3 as BADCAM this is the expected behaviour?
Are the nightly biases supposed to be used for R/Z at all? There are only `calibnight/biasnight-*` outputs/log-entries for B. Since this is the case generally and not just in this test_prod, I'd assume we currently don't use nightly biases for R/Z.

Re `make_dark_scripts`: it does put the bad exps inside the speclog, and I get 89 ZEROs and 26 DARKs for that night, of which 59 ZEROs (throwing out the first 30, but shouldn't this be 25 instead, else we're losing the first 5 ZEROs from the sequences as well, right?) and  6 DARKs of 300s (A,B,C,D + 2 daily).
Output files show COMMENT headers for exp. 111991++ for BIAS (i.e. for z3 only 5 nightly VCCD=on ZEROs are removed) and include 112013 also for the z3 DARK computation. I'll have a look what needs changing tomorrow. 

Given that, I updated the z3 DARKs last week before seeing this issue and including 20211203, so the bad 111941 has made it in the current DESI_SPECTRO_CALIB, will clean this up as well.

**Comment by Waelthus** [comment]
For my understanding (@akremin):
after merging  #1537 we would have a way to flag exposures in the exptabs and update the exposure list with the full set, right?
So in principle instead of creating the speclog (in `desispec.ccdcalib.make_dark_scripts`) I'd generate a merge of all zeros/darks in the exptables of all requested nights and could then check for `camera not in BADCAMWORD` and `LASTSTEP != 'ignore'` during `desi_dark_nonlinear`? I tried this approach yesterday, but it still failed because I didn't recognize the incomplete exptabs, yet...

**Comment by akremin** [comment]
Hi Michael, 

Thank you for your effort on this.

There is no need to wait on the referenced PR for recent nights like the one in question. The PR you refer to will update historical exposure tables so that we can do the same things on historical data.

And to answer another question you had -- you're correct, as of right now we only process the b cameras for nightly biases. 

The dark is selected by the pipeline -- so that is on me to select. You verifying that my code skips the first dark (marked as bad) is good to hear. 

What is currently ad-hoc and what Stephen would like you to work on is the selection of the zeros within the functions you are already referencing. He already uses the exposure tables, but what he doesn't do is read in or interpret the `BADCAMWORD` or `BADAMP` columns.

I think what you want to do is edit `desispec.ccdcalib._find_zeros()` to return either:
1.  a list of expids and a list of cameras, or 
2. more rigorously a dictionary where the keys are cameras and the values are a list of valid exposures.
I could see you trying two different things

Then edit `desispec.ccdcalib.compute_nightly_bias()` to use that information properly per-camera.

To generate the new outputs of `_find_zeros()`, I think you would start by:
1.  assuming any identified zeros that aren't in the exposure table are good for all cameras.
2. Throwing out any zeros in the exposure table with `BADCAMWORD!='' `and `BADAMPS!=''`. 
3. If there are enough other zeros, like the night in question, then proceed with those.
4. Else, translate the `CAMWORD`, `BADCAMWORD`, and `BADAMPS` into the valid cameras per exposure (I can supply code to do this)
5. Invert the problem so that you get the exposures that are valid for each camera.
6. Return that dictionary of (camera:expidlist) key value pairs.

If you don't want to do the inversion, you could also just find the cameras that are present in all exposures and return a single list of exposures with a list of cameras valid for all those exposures.


Let me know if you have questions and if you want the code snippet to translate the various column values into a list of cameras.

**Comment by Waelthus** [comment]
I gave this a try and think it is mostly working as it should right now for zeros (there's still some pieces missing, e.g. I guess I'll need to still add a check for CAMWORD, currently I'm assuming all requested cameras have been used), but I'll double check tomorrow and make this a PR.

For 20211203 where z3 zeros were bad I'm getting this (assuming I demand to use many zeros, else both would be cut to EXPs that have all cams working):
```
INFO:ccdcalib.py:468:compute_nightly_bias: Using 63 ZEROs for nightly bias 20211203 and cam z3
INFO:ccdcalib.py:468:compute_nightly_bias: Using 88 ZEROs for nightly bias 20211203 and cam z4
```

I think this should also address #1562 (as both BADAMPS and BADCAMWORD are checked), in the test on r1 for 20211217, I get:
```
INFO:ccdcalib.py:468:compute_nightly_bias: Using 25 ZEROs for nightly bias 20211217 and cam r1
INFO:ccdcalib.py:499:compute_nightly_bias: Rank 0 computing nightly bias for 20211217 r1
INFO:ccdcalib.py:239:compute_bias_file: read r1 images ...
INFO:ccdcalib.py:244:compute_bias_file: reading /global/cfs/cdirs/desi/spectro/data/20211217/00114409/desi-00114409.fits.fz
[lines with higher expid]
```

i.e. the bad 68341 though 68405 have not been used

**Comment by akremin** [comment]
This should be solved with PR #1570
